File: backend/src/server/handlers/transfer.rs
// 转存 API 处理器

use crate::server::AppState;
use crate::transfer::{TransferStatus, TransferTask};
use axum::{
    extract::{Path, State},
    http::StatusCode,
    response::Json,
};
use serde::{Deserialize, Serialize};
use tracing::{error, info};

/// API 响应结构
#[derive(Debug, Serialize)]
pub struct TransferApiResponse<T> {
    pub code: i32,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<T>,
}

impl<T: Serialize> TransferApiResponse<T> {
    pub fn success(data: T) -> Self {
        Self {
            code: 0,
            message: "success".to_string(),
            data: Some(data),
        }
    }

    pub fn error(code: i32, message: impl Into<String>) -> Self {
        Self {
            code,
            message: message.into(),
            data: None,
        }
    }
}

/// 业务响应码
pub mod error_codes {
    /// 需要提取码
    pub const NEED_PASSWORD: i32 = 1001;
    /// 提取码错误
    pub const INVALID_PASSWORD: i32 = 1002;
    /// 分享已失效
    pub const SHARE_EXPIRED: i32 = 1003;
    /// 分享不存在
    pub const SHARE_NOT_FOUND: i32 = 1004;
    /// 转存管理器未初始化
    pub const MANAGER_NOT_READY: i32 = 1005;
    /// 任务不存在
    pub const TASK_NOT_FOUND: i32 = 1006;
}

// ============================================
// 请求/响应结构
// ============================================

/// 创建转存任务请求
#[derive(Debug, Deserialize)]
pub struct CreateTransferRequest {
    /// 分享链接
    pub share_url: String,
    /// 提取码（可选）
    pub password: Option<String>,
    /// 网盘保存路径
    pub save_path: String,
    /// 网盘保存目录 fs_id
    pub save_fs_id: u64,
    /// 是否自动下载（不传使用全局配置）
    pub auto_download: Option<bool>,
    /// 本地下载路径（auto_download=true 时可选）
    pub local_download_path: Option<String>,
}

/// 创建转存任务响应
#[derive(Debug, Serialize)]
pub struct CreateTransferResponse {
    /// 任务 ID（创建成功时返回）
    pub task_id: Option<String>,
    /// 任务状态
    pub status: Option<TransferStatus>,
    /// 是否需要提取码
    pub need_password: bool,
}

/// 转存任务列表响应
#[derive(Debug, Serialize)]
pub struct TransferListResponse {
    pub tasks: Vec<TransferTask>,
    pub total: usize,
}

// ============================================
// API 处理器
// ============================================

/// POST /api/v1/transfers
/// 创建转存任务
pub async fn create_transfer(
    State(app_state): State<AppState>,
    Json(req): Json<CreateTransferRequest>,
) -> Json<TransferApiResponse<CreateTransferResponse>> {

    // 获取转存管理器
    let transfer_manager = {
        let guard = app_state.transfer_manager.read().await;
        match guard.clone() {
            Some(tm) => tm,
            None => {
                error!("转存管理器未初始化");
                return Json(TransferApiResponse::error(
                    error_codes::MANAGER_NOT_READY,
                    "转存管理器未初始化，请先登录",
                ));
            }
        }
    };

    // 创建转存请求
    let create_request = crate::transfer::manager::CreateTransferRequest {
        share_url: req.share_url,
        password: req.password,
        save_path: req.save_path,
        save_fs_id: req.save_fs_id,
        auto_download: req.auto_download,
        local_download_path: req.local_download_path,
    };

    // 创建任务
    match transfer_manager.create_task(create_request).await {
        Ok(response) => {
            if response.need_password {
                return Json(TransferApiResponse::error(
                    error_codes::NEED_PASSWORD,
                    response.error.unwrap_or_else(|| "需要提取码".to_string()),
                ));
            }

            if let Some(ref err) = response.error {
                // 根据错误内容返回不同的错误码
                let code = if err.contains("需要密码") || err.contains("需要提取码") {
                    error_codes::NEED_PASSWORD
                } else if err.contains("提取码错误") {
                    error_codes::INVALID_PASSWORD
                } else if err.contains("已失效") {
                    error_codes::SHARE_EXPIRED
                } else if err.contains("不存在") {
                    error_codes::SHARE_NOT_FOUND
                } else {
                    -1
                };

                return Json(TransferApiResponse::error(code, err.clone()));
            }

            info!("转存任务创建成功: task_id={:?}", response.task_id);
            Json(TransferApiResponse::success(CreateTransferResponse {
                task_id: response.task_id,
                status: response.status,
                need_password: false,
            }))
        }
        Err(e) => {
            let err_msg = e.to_string();

            error!("创建转存任务失败: {:?}", err_msg);

            // 根据错误内容返回不同的错误码
            let code = if err_msg.contains("提取码错误") {
                error_codes::INVALID_PASSWORD
            } else if err_msg.contains("已失效") {
                error_codes::SHARE_EXPIRED
            } else if err_msg.contains("不存在") {
                error_codes::SHARE_NOT_FOUND
            } else {
                -1
            };

            Json(TransferApiResponse::error(code, err_msg))
        }
    }
}

/// GET /api/v1/transfers
/// 获取所有转存任务
pub async fn get_all_transfers(
    State(app_state): State<AppState>,
) -> Result<Json<TransferApiResponse<TransferListResponse>>, StatusCode> {
    let transfer_manager = {
        let guard = app_state.transfer_manager.read().await;
        guard.clone().ok_or(StatusCode::SERVICE_UNAVAILABLE)?
    };

    let tasks = transfer_manager.get_all_tasks();
    let total = tasks.len();

    Ok(Json(TransferApiResponse::success(TransferListResponse {
        tasks,
        total,
    })))
}

/// GET /api/v1/transfers/:id
/// 获取单个转存任务
pub async fn get_transfer(
    State(app_state): State<AppState>,
    Path(task_id): Path<String>,
) -> Json<TransferApiResponse<TransferTask>> {
    let transfer_manager = {
        let guard = app_state.transfer_manager.read().await;
        match guard.clone() {
            Some(tm) => tm,
            None => {
                return Json(TransferApiResponse::error(
                    error_codes::MANAGER_NOT_READY,
                    "转存管理器未初始化",
                ));
            }
        }
    };

    match transfer_manager.get_task(&task_id).await {
        Some(task) => Json(TransferApiResponse::success(task)),
        None => Json(TransferApiResponse::error(
            error_codes::TASK_NOT_FOUND,
            "任务不存在",
        )),
    }
}

/// DELETE /api/v1/transfers/:id
/// 删除转存任务
pub async fn delete_transfer(
    State(app_state): State<AppState>,
    Path(task_id): Path<String>,
) -> Json<TransferApiResponse<String>> {
    info!("删除转存任务: {}", task_id);

    let transfer_manager = {
        let guard = app_state.transfer_manager.read().await;
        match guard.clone() {
            Some(tm) => tm,
            None => {
                return Json(TransferApiResponse::error(
                    error_codes::MANAGER_NOT_READY,
                    "转存管理器未初始化",
                ));
            }
        }
    };

    match transfer_manager.remove_task(&task_id) {
        Ok(()) => {
            info!("转存任务删除成功: {}", task_id);
            Json(TransferApiResponse::success("ok".to_string()))
        }
        Err(e) => {
            error!("删除转存任务失败: {:?}", e.to_string());
            Json(TransferApiResponse::error(-1, e.to_string()))
        }
    }
}

/// POST /api/v1/transfers/:id/cancel
/// 取消转存任务
pub async fn cancel_transfer(
    State(app_state): State<AppState>,
    Path(task_id): Path<String>,
) -> Json<TransferApiResponse<String>> {
    info!("取消转存任务: {}", task_id);

    let transfer_manager = {
        let guard = app_state.transfer_manager.read().await;
        match guard.clone() {
            Some(tm) => tm,
            None => {
                return Json(TransferApiResponse::error(
                    error_codes::MANAGER_NOT_READY,
                    "转存管理器未初始化",
                ));
            }
        }
    };

    match transfer_manager.cancel_task(&task_id) {
        Ok(()) => {
            info!("转存任务取消成功: {}", task_id);
            Json(TransferApiResponse::success("ok".to_string()))
        }
        Err(e) => {
            error!("取消转存任务失败: {:?}", e);
            Json(TransferApiResponse::error(-1, e.to_string()))
        }
    }
}

File: frontend/src/api/transfer.ts
import { apiClientWithErrorCode } from './client'
import { formatTimestampShort } from './utils'

const apiClient = apiClientWithErrorCode

// ============================================
// 业务错误码
// ============================================
export const TransferErrorCodes = {
  /** 需要提取码 */
  NEED_PASSWORD: 1001,
  /** 提取码错误 */
  INVALID_PASSWORD: 1002,
  /** 分享已失效 */
  SHARE_EXPIRED: 1003,
  /** 分享不存在 */
  SHARE_NOT_FOUND: 1004,
  /** 转存管理器未初始化 */
  MANAGER_NOT_READY: 1005,
  /** 任务不存在 */
  TASK_NOT_FOUND: 1006,
} as const

// ============================================
// 类型定义
// ============================================

/// 转存任务状态
export type TransferStatus =
    | 'queued'
    | 'checking_share'
    | 'transferring'
    | 'transfer_failed'
    | 'transferred'
    | 'downloading'
    | 'download_failed'
    | 'completed'

/// 分享页面信息
export interface SharePageInfo {
  shareid: string
  uk: string
  share_uk: string
  bdstoken: string
}

/// 分享文件信息
export interface SharedFileInfo {
  fs_id: number
  is_dir: boolean
  path: string
  size: number
  name: string
}

/// 转存任务
export interface TransferTask {
  id: string
  share_url: string
  password?: string
  save_path: string
  save_fs_id: number
  auto_download: boolean
  local_download_path?: string
  status: TransferStatus
  error?: string
  download_task_ids: string[]
  share_info?: SharePageInfo
  file_list: SharedFileInfo[]
  transferred_count: number
  total_count: number
  created_at: number
  updated_at: number
  failed_download_ids: string[]
  completed_download_ids: string[]
  download_started_at?: number
  /** 🔥 新增：转存文件名称（用于展示，从分享文件列表中提取） */
  file_name?: string
}

/// 创建转存任务请求
export interface CreateTransferRequest {
  share_url: string
  password?: string
  save_path: string
  save_fs_id: number
  auto_download?: boolean
  local_download_path?: string
}

/// 创建转存任务响应
export interface CreateTransferResponse {
  task_id?: string
  status?: TransferStatus
  need_password: boolean
}

/// 转存任务列表响应
export interface TransferListResponse {
  tasks: TransferTask[]
  total: number
}

/// 转存 API 错误
export interface TransferApiError {
  code: number
  message: string
  data?: any
}

// ============================================
// API 函数
// ============================================

/**
 * 创建转存任务
 * @throws TransferApiError 特殊错误（需要密码、密码错误等）
 */
export async function createTransfer(req: CreateTransferRequest): Promise<CreateTransferResponse> {
  return apiClient.post('/transfers', req)
}

/**
 * 获取所有转存任务
 */
export async function getAllTransfers(): Promise<TransferListResponse> {
  return apiClient.get('/transfers')
}

/**
 * 获取单个转存任务
 */
export async function getTransfer(taskId: string): Promise<TransferTask> {
  return apiClient.get(`/transfers/${taskId}`)
}

/**
 * 删除转存任务
 */
export async function deleteTransfer(taskId: string): Promise<string> {
  return apiClient.delete(`/transfers/${taskId}`)
}

/**
 * 取消转存任务
 */
export async function cancelTransfer(taskId: string): Promise<string> {
  return apiClient.post(`/transfers/${taskId}/cancel`)
}

// ============================================
// 辅助函数
// ============================================

/**
 * 获取状态文本
 */
export function getTransferStatusText(status: TransferStatus): string {
  const statusMap: Record<TransferStatus, string> = {
    queued: '排队中',
    checking_share: '检查分享...',
    transferring: '转存中',
    transfer_failed: '转存失败',
    transferred: '已转存',
    downloading: '下载中',
    download_failed: '下载失败',
    completed: '已完成',
  }
  return statusMap[status] || '未知'
}

/**
 * 获取状态类型（用于 Element Plus 组件）
 */
export function getTransferStatusType(status: TransferStatus): 'success' | 'warning' | 'danger' | 'info' {
  const typeMap: Record<TransferStatus, 'success' | 'warning' | 'danger' | 'info'> = {
    queued: 'info',
    checking_share: 'info',
    transferring: 'warning',
    transfer_failed: 'danger',
    transferred: 'success',
    downloading: 'warning',
    download_failed: 'danger',
    completed: 'success',
  }
  return typeMap[status] || 'info'
}

/**
 * 计算转存进度百分比
 */
export function calculateTransferProgress(task: TransferTask): number {
  if (task.total_count === 0) return 0
  return (task.transferred_count / task.total_count) * 100
}

/**
 * 判断是否为终止状态
 */
export function isTerminalStatus(status: TransferStatus): boolean {
  return ['transfer_failed', 'transferred', 'download_failed', 'completed'].includes(status)
}

/**
 * 判断错误码是否为需要密码
 */
export function isNeedPasswordError(error: TransferApiError): boolean {
  return error.code === TransferErrorCodes.NEED_PASSWORD
}

/**
 * 判断错误码是否为密码错误
 */
export function isInvalidPasswordError(error: TransferApiError): boolean {
  return error.code === TransferErrorCodes.INVALID_PASSWORD
}

/**
 * 获取简短的分享链接显示
 */
export function getShortShareUrl(url: string): string {
  // 从 URL 中提取关键部分
  const match = url.match(/\/s\/([a-zA-Z0-9_-]+)/)
  if (match) {
    return `pan.baidu.com/s/${match[1].substring(0, 8)}...`
  }
  // 回退：截断显示
  if (url.length > 40) {
    return url.substring(0, 37) + '...'
  }
  return url
}

/**
 * 格式化时间戳
 */
export const formatTransferTime = formatTimestampShort
